File: elixir/binary-search-tree/lib/binary_search_tree.ex
defmodule BinarySearchTree do
  @type bst_node :: %{data: any, left: bst_node | nil, right: bst_node | nil}

  @doc """
  Create a new Binary Search Tree with root's value as the given 'data'
  """
  @spec new(any) :: bst_node
  def new(data) do
    %{data: data, left: nil, right: nil}
  end

  @doc """
  Creates and inserts a node with its value as 'data' into the tree.
  """
  @spec insert(bst_node, any) :: bst_node
  def insert(tree, data)

  def insert(%{data: current, left: nil, right: right}, data) when data <= current do
    %{data: current, left: new(data), right: right}
  end

  def insert(%{data: current, left: left, right: right}, data) when data <= current do
    %{data: current, left: insert(left, data), right: right}
  end

  def insert(%{data: current, left: left, right: nil}, data) when data > current do
    %{data: current, left: left, right: new(data)}
  end

  def insert(%{data: current, left: left, right: right}, data) when data > current do
    %{data: current, left: left, right: insert(right, data)}
  end

  @doc """
  Traverses the Binary Search Tree in order and returns a list of each node's data.
  """
  @spec in_order(bst_node) :: [any]
  def in_order(tree)

  def in_order(nil), do: []

  def in_order(%{data: data, left: left, right: right}) do
    in_order(left)
    |> Enum.concat([data])
    |> Enum.concat(in_order(right))
  end
end

File: go/binary-search-tree/binary_search_tree.go
package binarysearchtree

// BinarySearchTree is tree structure that stores values in a naturally ordered manner, but does not have
// any type of balancing between the sides.
type BinarySearchTree struct {
	left  *BinarySearchTree
	data  int
	right *BinarySearchTree
}

// NewBst creates and returns a new BinarySearchTree.
func NewBst(i int) *BinarySearchTree {
	return &BinarySearchTree{data: i, left: nil, right: nil}
}

// Insert inserts an int into the BinarySearchTree.
// Inserts happen based on the rules of a binary search tree
func (bst *BinarySearchTree) Insert(i int) {
	current := bst
	for {
		goesLeft := i <= current.data
		if goesLeft && current.left == nil {
			current.left = NewBst(i)
			return
		}
		if !goesLeft && current.right == nil {
			current.right = NewBst(i)
			return
		}
		if goesLeft {
			current = current.left
		} else {
			current = current.right
		}
	}

}

// SortedData returns the ordered contents of BinarySearchTree as an []int.
// The values are in increasing order starting with the lowest int value.
// A BinarySearchTree that has the numbers [1,3,7,5] added will return the
// []int [1,3,5,7].
func (bst *BinarySearchTree) SortedData() []int {
	stack := []*BinarySearchTree{}
	sorted := []int{}

	for current := bst; current != nil || !isEmpty(&stack); {
		for current != nil {
			push(&stack, current)
			current = current.left
		}
		current, _ = pop(&stack)
		push(&sorted, current.data)
		current = current.right
	}

	return sorted
}
